### PR TITLE
feat: allow renaming columns of duckdb

### PR DESCRIPTION
**Pull request by sebffischer**
We rename by creating a new view from the given table / view with renamed columns + update the `self$table` and `self$primary_key` fields. Maybe it is a little inconsistent to not have this for the other data backends. Should I do this as well?

Can we somehow avoid that this function is called on a backend that is part of a task?

### CODE REVIEW

**Comment by mllg** [comment]
> Can we somehow avoid that this function is called on a backend that is part of a task?

We usually treat backends read-only. You could ensure this by making the new column names an argument of the constructor.

**Comment by sebffischer** [comment]
I think it is fine. I mean it is documented that one should not do it.

**Comment by sebffischer** [comment]
Maybe it is a little weird, that `Task$rename()` does not call `backend$rename()`. 
Maybe it should try doing so and if it is not available revert to `DataBackendRename` ?

**Comment by sebffischer** [comment]
@mllg ?

**Comment by sebffischer** [comment]
@mllg

**Comment by sebffischer** [comment]
this also fixes the bug that the `...` was not passed on to the duckdb constructor

File: R/DataBackendDuckDB.R [hunk 1]
```diff
@@ -71,6 +75,10 @@ DataBackendDuckDB = R6Class("DataBackendDuckDB", inherit = DataBackend, cloneabl
         self$levels = self$distinct(rows = NULL, cols = strings_as_factors)
       }
 
+      if (rename) {
+        private$.rename()
```
Comment (mllg) [inline R/DataBackendDuckDB.R:79]: should not be a function, only called once.

File: R/DataBackendDuckDB.R [hunk 2]
```diff
@@ -259,6 +267,29 @@ DataBackendDuckDB = R6Class("DataBackendDuckDB", inherit = DataBackend, cloneabl
...
+    .rename = function() {
+      old = self$colnames
+      new = make.names(old, unique = TRUE)
+
+      if (all(new == old)) {
+        return(invisible(self))
+      }
+
+      existing_tables = DBI::dbGetQuery(private$.data, "PRAGMA show_tables")$name
+      table_new = make.unique(c(existing_tables, self$table), sep = "_")[length(existing_tables) + 1L]
```
Comment (mllg) [inline R/DataBackendDuckDB.R:280]: add a comment

File: R/DataBackendDuckDB.R [hunk 3]
```diff
@@ -259,6 +267,29 @@ DataBackendDuckDB = R6Class("DataBackendDuckDB", inherit = DataBackend, cloneabl
...
+    },
+    .rename = function() {
+      old = self$colnames
+      new = make.names(old, unique = TRUE)
+
+      if (all(new == old)) {
+        return(invisible(self))
+      }
+
+      existing_tables = DBI::dbGetQuery(private$.data, "PRAGMA show_tables")$name
```
Comment (sebffischer) [inline R/DataBackendDuckDB.R:279]: comment what is done here

File: R/DataBackendDuckDB.R [hunk 4]
```diff
@@ -259,6 +267,29 @@ DataBackendDuckDB = R6Class("DataBackendDuckDB", inherit = DataBackend, cloneabl
...
+      table_new = make.unique(c(existing_tables, self$table), sep = "_")[length(existing_tables) + 1L]
+      primary_key_new = new[old == self$primary_key]
```
Comment (mllg) [inline R/DataBackendDuckDB.R:281]: only used in l. 290

File: tests/testthat/test_duckdb.R [hunk 5]
```diff
@@ -62,3 +62,9 @@ test_that("multiple parquet file", {
   b = as_duckdb_backend(files)
   expect_backend(b)
 })
+
+test_that("rename works", {
+  file = system.file(file.path("extdata", "weird_names.parquet"), package = "mlr3db")
+  brn = as_duckdb_backend(file, rename = TRUE)
+  expect_true(brn$colnames[1L] == "Sepal.Length")
```
Comment (mllg) [inline tests/testthat/test_duckdb.R:69]: add some more tests!

**Review by mllg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**